src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+                _ => throw new NotSupportedException()
+            };
+
+            string metadataInitSource = @$"_{typeFriendlyName} = {collectionTypeInfoValue};";
+            return GenerateForType(typeMetadata, metadataInitSource);
+        }
+
+        private string GenerateForObject(TypeMetadata typeMetadata)
+        {
+            string typeCompilableName = typeMetadata.CompilableName;
+            string typeFriendlyName = typeMetadata.FriendlyName;
+
+            string createObjectFuncTypeArg = typeMetadata.ConstructionStrategy == ObjectConstructionStrategy.ParameterlessConstructor
+                ? $"createObjectFunc: static () => new {typeMetadata.CompilableName}()"
+                : "createObjectFunc: null";
+
+            List<PropertyMetadata>? properties = typeMetadata.PropertiesMetadata;
+
+            StringBuilder sb = new();
+
+            sb.Append($@"JsonTypeInfo<{typeCompilableName}> objectInfo = {JsonMetadataServicesClassName}.CreateObjectInfo<{typeCompilableName}>();
+                    _{typeFriendlyName} = objectInfo;
+");
+
+            string propInitFuncVarName = $"{typeFriendlyName}{PropInitFuncVarName}";
+
+            sb.Append($@"
+                    {JsonMetadataServicesClassName}.InitializeObjectInfo(
+                        objectInfo,
+                        {OptionsInstanceVariableName},
+                        {createObjectFuncTypeArg},
+                        {propInitFuncVarName},
+                        {GetNumberHandlingAsStr(typeMetadata.NumberHandling)});");
+
+            string metadataInitSource = sb.ToString();
+            string? propInitFuncSource = GeneratePropMetadataInitFunc(typeMetadata.IsValueType, propInitFuncVarName, properties);
+
+            return GenerateForType(typeMetadata, metadataInitSource, propInitFuncSource);
+        }
+
+        private string GeneratePropMetadataInitFunc(
+            bool declaringTypeIsValueType,
+            string propInitFuncVarName,
+            List<PropertyMetadata>? properties)
+        {
+            const string PropVarName = "properties";
+            const string JsonContextVarName = "jsonContext";
+            const string JsonPropertyInfoTypeName = "JsonPropertyInfo";
+
+            string propertyArrayInstantiationValue = properties == null
+                ? $"System.Array.Empty<{JsonPropertyInfoTypeName}>()"
+                : $"new {JsonPropertyInfoTypeName}[{properties.Count}]";
+
+            StringBuilder sb = new();
+
+            sb.Append($@"private static {JsonPropertyInfoTypeName}[] {propInitFuncVarName}(JsonSerializerContext context)
+        {{
```

Comment:
Double brace intentional?

Reply:
Yes, it's to distinguish between a "hole" in the interpolated string and an actual brace which I am writing to the generated source file.